tests/integration/test_external_url.py

```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python3
+# Copyright 2021 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Test various aspects of `external_url`.
+
+1. When external_url is set (with path prefix) via traefik, self-scraping job is 'up'.
+2. When external_url is set (with path prefix) via config option to a different value,
+   self-scraping job is 'up'.
+"""
+
+import asyncio
+import logging
+import subprocess
+import urllib.request
+
+import pytest
+from helpers import oci_image
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+prometheus_app_name = "prometheus"
+prometheus_resources = {"prometheus-image": oci_image("./metadata.yaml", "prometheus-image")}
+
+# Two prometheus units are sufficient to test potential interactions between multi-unit
+# deployments and external_url
+num_units = 2
+
+# The period of time required to be idle before `wait_for_idle` returns is set to 90 sec because
+# the default scrape_interval in prometheus is 1m.
+idle_period = 90
+
+# Set up microk8s metallb addon, needed by traefik
+cmd = [
+    "sh",
+    "-c",
+    "ip -4 -j route | jq -r '.[] | select(.dst | contains(\"default\")) | .prefsrc'",
+]
+result = subprocess.run(cmd, check=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+ip = result.stdout.decode("utf-8").strip()
+
+cmd = ["sg", "microk8s", "-c", f"microk8s enable metallb:{ip}-{ip}"]
+subprocess.run(cmd, check=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+
+async def test_setup_env(ops_test: OpsTest):
+    await ops_test.model.set_config(
+        {"logging-config": "<root>=WARNING; unit=DEBUG", "update-status-hook-interval": "60m"}
+    )
+
+
+@pytest.mark.abort_on_fail
+async def test_self_scrape_job_is_up_with_traefik(ops_test: OpsTest, prometheus_charm):
+    await asyncio.gather(
+        ops_test.model.deploy(
+            prometheus_charm,
+            resources=prometheus_resources,
+            application_name=prometheus_app_name,
+            num_units=num_units,
+            trust=True,
+        ),
+        ops_test.model.deploy(
+            "ch:traefik-k8s",
+            application_name="traefik",
+            channel="edge",
+        ),
+    )
+
+    app_names = [prometheus_app_name, "traefik"]
+
+    await asyncio.gather(
+        # ops_test.model.applications["traefik"].set_config(
+        #     {"alert-rules-path": "src/with_extra_alert_rule"}
+        # )
+        ops_test.model.add_relation(f"{prometheus_app_name}:ingress", "traefik"),
+        ops_test.model.wait_for_idle(
+            apps=app_names,
+            status="active",
+            wait_for_units=num_units,
+            timeout=600,
+            idle_period=idle_period,
+        ),
+    )
```

Comment:
Shouldn't these be run sequentially rather than in parallel? When run in parallel, `wait_for_idle` could in theory return before `add_relation` starts running.

Reply:
Done.